File: coding/src/com/algos/li10_tree/bst/revised/L98_ValidateBinarySearchTree.java
package  com.algos.li10_tree.bst.revised;

import  com.algos.li30_model.TreeNode;

public class L98_ValidateBinarySearchTree {
    public static void main(String[] args) {
        L98_ValidateBinarySearchTree obj = new L98_ValidateBinarySearchTree();

        L701_InsertIntoABinarySearchTree bst = new L701_InsertIntoABinarySearchTree();
        TreeNode root = bst.insertIntoBST(null, 10);
		bst.insertIntoBST(root, 2);
		bst.insertIntoBST(root, 4);
		bst.insertIntoBST(root, 6);
		bst.insertIntoBST(root, 5);
		bst.insertIntoBST(root, 15);
		bst.insertIntoBST(root, 3);

        System.out.println(obj.isValidBST(root));
    }

    // Problem: https://leetcode.com/problems/validate-binary-search-tree/
    // Idea: Pass min and max to all the recursive calls so that all child nodes should be in that range.
    // Because, immediate child might be in order but grand childs might not be in order.
    // Time: O(n)
    // space: O(1)
    public boolean isValidBST(TreeNode root) {
        return isBst(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    private boolean isBst(TreeNode root, long min, long max){
        if(root == null)
            return true;

        return (root.val > min && root.val < max) && isBst(root.left, min, root.val) && isBst(root.right, root.val, max);
    }
}

File: dsa/patterns/10_trees/bst/python/l98__validate_binary_search_tree.py
# Tree / BST
# Problem: https://leetcode.com/problems/validate-binary-search-tree/
# Idea:
# 1. Pass min and max values to each recursive call.
# 2. Each node must be in range (min, max).
# 3. Immediate child might satisfy BST but grandchild could violate it.
# Time: O(n)
# Space: O(h) recursion stack

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_valid_bst(root):
    def helper(node, min_val, max_val):
        if not node:
            return True
        if not (min_val < node.val < max_val):
            return False
        return helper(node.left, min_val, node.val) and helper(node.right, node.val, max_val)
    
    return helper(root, float('-inf'), float('inf'))


# -------- Test --------
# Helper to insert into BST
def insert_into_bst(root, val):
    if not root:
        return TreeNode(val)
    if val < root.val:
        root.left = insert_into_bst(root.left, val)
    else:
        root.right = insert_into_bst(root.right, val)
    return root

# Construct BST
root = None
for val in [10, 2, 4, 6, 5, 15, 3]:
    root = insert_into_bst(root, val)

print(is_valid_bst(root))  # Expected: False (this tree violates BST rules)
